lib/new_relic/agent/log_event_aggregator.rb

```diff
@@ -0,0 +1,185 @@
+      #
+      # We have to keep the aggregated payloads in a separate shape, though, to
+      # work with the priority sampling buffers
+      def self.payload_to_melt_format(data)
+        metadata, items = data
+        payload = [{
+          common: { attributes: metadata[:linking] },
+          logs: items.map(&:last)
+        }]
+
+        return [payload, items.size]
+      end
+
+      def harvest!
+        record_customer_metrics()
+        super
+      end
+
+      def reset!
+        @counter_lock.synchronize do
+          @seen = 0
+          @seen_by_severity.clear
+        end
+        super
+      end
+
+      private
+
+      # Slightly awkward, but we want the base harvest but without
+      # this method altering the metadata we gather...
+      def reservoir_metadata metadata
+        metadata
+      end
+
+      def register_capacity_callback
+        NewRelic::Agent.config.register_callback(self.class.capacity_key) do |max_samples|
+          NewRelic::Agent.logger.debug "#{self.class.named} max_samples set to #{max_samples}"
+          @lock.synchronize do
+            @buffer.capacity = max_samples
+          end
+        end
+      end
+
+      def after_harvest metadata
+        linking = { PLUGIN_TYPE_KEY => PLUGIN_TYPE }
+        LinkingMetadata.append_service_linking_metadata(linking)
+        metadata[:linking] = linking
+
+        dropped_count = metadata[:seen] - metadata[:captured]
+        note_dropped_events(metadata[:seen], dropped_count)
+        record_supportability_metrics(metadata[:seen], metadata[:captured], dropped_count)
+      end
+
+      # To avoid paying the cost of metric recording on every line, we hold
+      # these until harvest before recording them
+      def record_customer_metrics
+        @counter_lock.synchronize do
+          return unless @seen > 0
+
+          NewRelic::Agent.increment_metric(LINES, @seen)
+          @seen_by_severity.each do |(severity, count)|
+            NewRelic::Agent.increment_metric(line_metric_name_by_severity(severity), count)
+          end
+
+          @seen = 0
+          @seen_by_severity.clear
+        end
+      end
+
+      def line_metric_name_by_severity(severity)
+        @line_metrics ||= {}
+        @line_metrics[severity] ||= "Logging/lines/#{severity}".freeze
+      end
+
+      def note_dropped_events total_count, dropped_count
+        if dropped_count > 0
+          NewRelic::Agent.logger.warn("Dropped #{dropped_count} log events out of #{total_count}.")
+        end
+      end
+
+      def record_supportability_metrics total_count, captured_count, dropped_count
+        return unless total_count > 0
+
+        NewRelic::Agent.increment_metric("Supportability/Logging/Customer/Seen", total_count)
+        NewRelic::Agent.increment_metric("Supportability/Logging/Customer/Sent", captured_count)
+        NewRelic::Agent.increment_metric("Supportability/Logging/Customer/Dropped", dropped_count)
```

Comment:
TODO: Update names once finalized. Also potentially record customer-facing metric for dropped lines.